server/datastore/cached_mysql/cached_mysql.go

```diff
@@ -2,51 +2,113 @@ package cached_mysql
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
+	"reflect"
 	"time"
 
 	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jinzhu/copier"
 	"github.com/patrickmn/go-cache"
 )
 
+const (
+	appConfigKey                      = "AppConfig:%s"
+	defaultAppConfigExpiration        = 1 * time.Second
+	packsHostKey                      = "Packs:host:%d"
+	defaultPacksExpiration            = 1 * time.Minute
+	scheduledQueriesKey               = "ScheduledQueries:pack:%d"
+	defaultScheduledQueriesExpiration = 1 * time.Minute
+	teamAgentOptionsKey               = "TeamAgentOptions:team:%d"
+	defaultTeamAgentOptionsExpiration = 1 * time.Minute
+)
+
+// Cloner represents any type that can clone itself. Used by types to provide a more efficient clone method.
+type Cloner interface {
+	Clone() (interface{}, error)
+}
+
+func clone(v interface{}) (interface{}, error) {
+	if cloner, ok := v.(Cloner); ok {
+		return cloner.Clone()
+	}
+
+	vv := reflect.ValueOf(v)
+	isPtr := false
+	if vv.Kind() == reflect.Ptr {
+		isPtr = true
+		vv = vv.Elem()
+	}
+
+	clone := reflect.New(vv.Type())
+	err := copier.Copy(clone.Interface(), v)
+	if err != nil {
+		return nil, err
+	}
+
+	if isPtr {
+		return clone.Interface(), nil
+	}
+
+	return clone.Elem().Interface(), nil
```

Comment:
Should we just force implementation of `Cloner` for values that we want to store in that cache? Given that we cherry-pick those, it doesn't seem so unreasonable, and we wouldn't have to add the third-party dependency for the generic case (assuming this is just used for that).

Reply:
Discussed this with @zwass a little bit. I added this implementation after learning the intention behind cloning `AppConfig` in this [review comment](https://github.com/fleetdm/fleet/pull/3071#discussion_r754631585) and that we were currently *not* cloning all values that were being returned. For example, `(*cachedMysql).ListPacksForHost` does not clone the slice returned.

> Given that we cherry-pick those

By wrapping the cache and cloning all values returned, we can ensure that we are *always* cloning. It also means less code to add caching for another method. I don't think we want to be cherry-picking which cached methods clone and which do not.

Reply:
Obviously this does not solve the problem of clearing or updating the cache for datastore methods that modify resources, but it's still an improvement IMO.

Reply:
> I don't think we want to be cherry-picking which cached methods clone and which do not.

I think you misunderstood my comment, I meant that we are cherry-picking what values are actually cached, not cloned (i.e. those that have keys defined here: https://github.com/fleetdm/fleet/pull/4193/files#diff-81dff86684050db7d9acc3fc4949e530942a88fe79b05d0aaf25c3db546efeacR16-R23)

But anyway I left this comment before seeing how AppConfig.Clone was implemented (basically, using copier, not with a custom-made clone) and noticing that some other cached values were not currently cloned. Given that, I think it's fine to keep it as-is with the generic reflection-based fallback.